fileserver.gr

```diff
@@ -8,59 +8,83 @@ import String from "string"
 import Mediatype from "./lib/mediatype"
 import Stringutil from "./lib/stringutil"
 
-let serve = (abs_path) => {
-    // Trim the leading /
-    let path = String.slice(1, String.length(abs_path), abs_path)
-    File.fdWrite(File.stderr, "Fileserver: Loading file ")
-    File.fdWrite(File.stderr, path)
-    File.fdWrite(File.stderr, "\n")
+let internalError = () => {
+  // Is stdout correct here?
+  let result = File.fdWrite(File.stdout, "Status: 500\n\nInternal Server Error")
+  match (result) {
+    Err(err) => {
+      // TODO: What do you do on a write error?
+      void
+    },
+    Ok(_nBytes) => void,
+  }
+}
 
-    // Open file
-    // TODO: This throws an error when something goes wrong.
-    // But Grain 0.3 does not have a try/catch yet.
-    // When we figure out how, we need to catch the error here
-    // and do a 404 or 500.
-    let input = File.pathOpen(
-        File.pwdfd,
-        [],
-        path,
-        [],
-        [File.FdRead],
-        [],
-        [],   
-    )
+let notFound = () => {
+  let result = File.fdWrite(File.stdout, "Status: 404\n\nNot Found")
+  match (result) {
+    Err(err) => {
+      // TODO: What do you do on a write error?
+      void
+    },
+    Ok(_nBytes) => void,
+  }
+}
 
-    File.fdWrite(File.stdout, "Content-Type: ")
-    File.fdWrite(File.stdout, Mediatype.guess(path))
-    File.fdWrite(File.stdout, "\n\n")
-    
-    // Pipe output to STDOUT
-    let rec pipe = (in, out) => {
-        let (d, len) = File.fdRead(in, 1024)
-        File.fdWrite(out, d)
-        if (len > 0) {
-            pipe(in, out)
-        }
-    }
-    pipe(input, File.stdout)
-    File.fdClose(input)
+// Pipe output to STDOUT
+let rec pipe = (in, out) => {
+  let res = File.fdRead(in, 1024)
+  match (res) {
+    Err(err) => {
+      // TODO: What to do on a read error?
```

Comment:
Just thinking out loud here... and I may be totally misguided....

Maybe we should return an `Err` option as soon as an error is encountered, and otherwise keep recursing until the input has been exhausted, then return `Ok(len)` or something.

That way, we would be able to handle (somewhat gracefully) the situation where a `pipe` got interrupted. I'm not sure that in practice this helps us much at this point... but it feels sensible.

Reply:
I like it!